File: data-structures/1_linked_list/1_singlyLinkedList/singlyLinkedList.ts
interface INode<T> {
  value: T;
  next: INode<T> | null;
}

class SingleLinkedList<T> {
  head: INode<T> | null = null;

  generateNode(value: T): INode<T> {
    return {
      value,
      next: null,
    };
  }

  traverse() {
    let currentNode = this.head;
    while (currentNode) {
      console.log(currentNode.value);
      currentNode = currentNode.next;
    }
  }

  append(value: T) {
    const newNode = this.generateNode(value);
    if (!this.head) {
      this.head = newNode;
      return;
    }

    let currentNode = this.head;
    while (currentNode.next) {
      currentNode = currentNode.next;
    }

    //Append to the last node
    currentNode.next = newNode;
  }

  insertAfter(targetVal: T, newValue: T) {
    let currentNode = this.head;
    while (currentNode) {
      if (currentNode.value === targetVal) {
        const newNode = this.generateNode(newValue);
        newNode.next = currentNode.next;
        currentNode.next = newNode;
        return;
      }
      currentNode = currentNode.next;
    }
  }

  delete(value: T) {
    let prevNode: INode<T> | null = null;
    let currentNode = this.head;

    while (currentNode) {
      if (currentNode.value === value) {
        // If the node is head
        if (!prevNode) {
          this.head = currentNode.next;
          return;
        }
        prevNode.next = currentNode.next;
        return;
      }
      prevNode = currentNode;
      currentNode = currentNode.next;
    }
  }
}

const linkedList = new SingleLinkedList();
linkedList.append(5);
linkedList.append(6);
linkedList.append(7);
linkedList.append(8);
// linkedList.traverse();

// linkedList.insertAfter(6, 88);
// linkedList.traverse();

//Removing head
linkedList.delete(5);
//Remove some node in middle
linkedList.delete(7);
linkedList.traverse();

File: data-structures/1_linked_list/1_singlyLinkedList/singlyLinkedList.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def append(self, value):
        newNode = Node(value)
        if(self.head is None):
            self.head = newNode
            return
        
        currentNode = self.head
        while(currentNode.next):
            currentNode = currentNode.next
        
        currentNode.next = newNode
    
    def traverse(self):
        currentNode = self.head
        while(currentNode):
            print(currentNode.value)
            currentNode = currentNode.next

    def insertAfter(self, targetValue, newValue):
        currentNode = self.head
        while(currentNode):
            if(currentNode.value == targetValue):
                newNode = Node(newValue)
                newNode.next = currentNode.next
                currentNode.next = newNode
                return
            currentNode = currentNode.next

    def delete(self, targetValue):
        prevNode = None
        currentNode = self.head
        while(currentNode):
            if(currentNode.value == targetValue ):
                # If target node is head 
                if(prevNode is None):
                    self.head = currentNode.next
                    return
                
                prevNode.next = currentNode.next
                return
            prevNode = currentNode
            currentNode = currentNode.next



ln = LinkedList()
ln.append(5)
ln.append(6)
ln.append(7)
ln.append(8)
ln.insertAfter(6, 88)
ln.delete(8)
ln.traverse()
        
